annnufan/assignment1/solution_1.cpp

```diff
@@ -16,12 +16,14 @@ bool word_anagram(std::string word_1, std::string word_2, bool case_sens) {
 	}
 
 	std::vector<int> count(256, 0);
+	
+	//String used signed char. So, it needs to add 127 for each char for counting sort 

```

Comment:
There are many approaches of using `auto`. In my opinion, we should use it when it makes the code shorter but does not "hide" information from the user.
In my opinion, using `char` (or `unsigned char`) explicitly (as in `for (char ch : word_1)`) is more readable, especially if the reader is not sure whether it is signed or unsigned. (In general I think that `auto` is not a good replacement for primitive types).

Also:
string operator[] does not return a char, but a char reference! (`char&`)
This means that if you change `ch` in the loop, it will modify the string. This can be a source of bugs!
consider the following code:
```C++
for (auto ch : word) {  // Equivalent to: for (char& ch : word)
    ch += 127;
    count[ch]++;
}
```
This will change every character of the string, because `ch` is a reference to the character in the string! A bug that can be hard to chase down, especially in a large system.
Using `char` explicitly solves this.